pkg/sql/doctor/doctor.go

```diff
@@ -46,32 +56,121 @@ func NewDescGetter(rows []DescriptorTableRow) (catalog.MapDescGetter, error) {
+	namespaceTable []NamespaceTableRow,
+	verbose bool,
+	stdout io.Writer,
+) (ok bool, err error) {
+	fmt.Fprintf(
+		stdout, "Examining %d descriptors and %d namespace entries...\n",
+		len(descTable), len(namespaceTable))
+	descGetter, err := newDescGetter(descTable)
 	if err != nil {
 		return false, err
 	}
+	nMap := newNamespaceMap(namespaceTable)
+
 	var problemsFound bool
 	for _, row := range descTable {
+		desc, ok := descGetter[descpb.ID(row.ID)]
+		if !ok {
+			// This should never happen as ids are parsed and inserted from descTable.
+			log.Fatalf(context.Background(), "Descriptor id %d not found", row.ID)
+		}
+
+		if int64(desc.GetID()) != row.ID {
+			fmt.Fprint(stdout, reportMsg(desc, "different id in the descriptor: %d", row.ID))
+			problemsFound = true
+			continue
+		}
+
 		table, isTable := descGetter[descpb.ID(row.ID)].(catalog.TableDescriptor)
-		if !isTable {
-			// So far we only examine table descriptors. We may add checks for other
-			// descriptors in later versions.
+		if isTable {
+			if err := table.Validate(context.Background(), descGetter); err != nil {
+				problemsFound = true
+				fmt.Fprint(stdout, reportMsg(desc, " %s", err))
+			}
+		} else {
+			_, isDatabase := descGetter[descpb.ID(row.ID)].(catalog.DatabaseDescriptor)
+			if !isDatabase {
+				// So far we only examine table and database descriptors. We may add
+				// checks for other descriptors in later versions.
+				continue
+			}
+		}
+
+		// Process namespace entries point1ing to this descriptor.
+		names, ok := nMap[row.ID]
+		if !ok {
+			if !desc.Dropped() {
+				fmt.Fprint(stdout, reportMsg(desc, "not being dropped but no namespace entry found"))
+				problemsFound = true
+			}
 			continue
 		}
-		if int64(table.GetID()) != row.ID {
-			fmt.Fprintf(stdout, "Table %3d: different id in the descriptor: %d", row.ID, table.GetID())
+		// We delete all pointed descriptors to leave what is missing in the
+		// descriptor table.
+		delete(nMap, row.ID)
+
+		var found bool
+		for _, n := range names {
+			if n.Name == desc.GetName() &&
+				n.ParentSchemaID == desc.GetParentSchemaID() &&
+				n.ParentID == desc.GetParentID() {
+				found = true
+				break
+			}
+		}
+		if !found {
+			fmt.Fprint(stdout, reportMsg(desc, "could not find name in namespace table"))
 			problemsFound = true
 			continue
 		}
-		if err := table.Validate(context.Background(), descGetter); err != nil {
+		if verbose {
+			fmt.Fprint(stdout, reportMsg(desc, "processed"))
+		}
+	}
+
+	// Now go over all namespace entries that don't point to descriptors in the
+	// descriptor table.
+	for id, ni := range nMap {
+		if id == 29 {
+			continue
+		}
+		if descpb.ID(id) == descpb.InvalidID {
+			fmt.Fprintf(stdout, "Row(s) %+v: NULL value found\n", ni)
 			problemsFound = true
-			fmt.Fprintf(stdout, "Table %3d: %s\n", table.GetID(), err)
-		} else if verbose {
-			fmt.Fprintf(stdout, "Table %3d: validated\n", table.GetID())
+			continue
 		}
+		fmt.Fprintf(
```

Comment:
Temporary schemas will have namespace entries but not descriptors (for now). You can handle this by checking `strings.HasPrefix(name, "pg_temp_")`